vmdk_plugin/refcnt.go

```diff
@@ -192,23 +234,72 @@ func (r refCountsMap) decr(vol string) (uint, error) {
 		// we should NEVER get here. Even if Docker sends Unmount before Mount,
 		// it should be caught in previous check. So delete the entry (in case
 		// someone upstairs does 'recover', and panic.
-		delete(r, vol)
+		delete(r.refMap, vol)
 		log.Panicf("decr: refcnt corruption (rc.count=0), name=%s", vol)
 	}
 
 	rc.count--
-	if rc.count == 0 {
-		defer delete(r, vol)
+	// Deletes the refcount only if there are no references and no lock holders
+	if rc.count == 0 && rc.lockers == 0 {
+		delete(r.refMap, vol)
 	}
 	return rc.count, nil
 }
 
+func (r refCountsMap) lockVolume(vol string) {
+	// Lock the refCountsMap
+	r.mtx.Lock()
+
+	rc := r.refMap[vol]
+	if rc == nil {
+		rc = newRefCount()
+		r.refMap[vol] = rc
+	}
+
+	// Increment the lockers while the refCountsMap is still locked
+	rc.lockers++
+
+	// Unlock the refCountsMap
+	r.mtx.Unlock()
+
+	// Locks the volume outside the refCountsMap
+	// Mutex so we dont block everyone else
+	rc.mtx.Lock()
+}
+
+func (r refCountsMap) unlockVolume(vol string) error {
+	// Lock the refCountsMap
+	r.mtx.Lock()
+	defer r.mtx.Unlock()
+
+	rc := r.refMap[vol]
+	if rc == nil {
+		return fmt.Errorf("Volume %s has no refcount, cannot be unlocked", vol)
+	}
+
+	if rc.lockers == 0 {
+		return fmt.Errorf("Volume %s has no lockers, cannot be unlocked", vol)
+	}
+
+	rc.lockers--
+	rc.mtx.Unlock()
+
+	// Deletes the refcount only if there are no references and no lock holders
+	if rc.count == 0 && rc.lockers == 0 {
+		delete(r.refMap, vol)
+	}
+	return nil
+}
+
 // enumberates volumes and  builds refCountsMap, then sync with mount info
 func (r refCountsMap) discoverAndSync(c *client.Client, d *vmdkDriver) error {
 	// we assume to  have empty refcounts. Let's enforce
-	for name := range r {
-		delete(r, name)
+
+	r.mtx.Lock() // Lock the refCountsMap to purge the refcounts
+	for name := range r.refMap {
+		delete(r.refMap, name)
 	}
+	r.mtx.Unlock() // Unlock.
 
```

Comment:
With the current code no. I have protected those sections for future usage, like online recovery. If thats unlikely i can remove them (but they don't hurt anyone :)